### DISCUSSION TITLE
Monotonic Ironing?

### DISCUSSION THREAD

**Discussion by ViralWick** [smartavionics/Cura#108] [Ideas]
Hey Mark! Just wanted to write to first thank you for your effort and contribution to the Cura community! I just downloaded your latest release because I was curious about Monotonic skin lines, and I can tell it works flawlessly! 
However, based on my tests this setting does not affect the travel moves when it comes to ironing. It would be really cool to see that implemented when ironing is enabled, as this would also contribute to the quality of the top layer! 

Here I have monotonic skin lines and ironing both enabled, and you can see how while ironing, it skips a patch, comes back to it and changed direction when going over it which creates a visible difference in sheen on the top layer!

My guess is that it has to do with the fact that monotonic skin lines only wok with lines pattern and the ironing pattern doesn't have a lines option?

https://user-images.githubusercontent.com/84275232/118398180-1d529880-b660-11eb-8c60-88c4d4cfe683.mp4

I really hope this gets implemented in the official Cura release soon as well! I'm surprised they haven't prioritized it...

**Comment by smartavionics** [comment]
Hi @ViralWick , thanks for the kind words. You're right, ironed lines are not monotonic because they can't (currently) be printed with lines pattern rather thank zig-zag and my monotonic lines code won't work with zig-zag. So I have made the changes required for ironed lines to use lines pattern and can make you a build so you can test that. What system are you using?

Unfortunately, I no longer submit any new features to UM for inclusion in their releases so unless someone else does it, you won't see monotonic skin lines in a UM release.

**Reply by Sophist-UK** [reply]
> Unfortunately, I no longer submit any new features to UM for inclusion in their releases so unless someone else does it, you won't see monotonic skin lines in a UM release.

I assume that you no longer submit improvements because you started to believe it was a waste of time - because Ultimaker were unreceptive to them. Unfortunately this is what often seems to happen when a commercial company takes hold of an open source project - any promises of continuing to support the open source community soon disappear in favour of their own proprietary interests and the community starts to be ignored.

All the community can do, I think, is to rally behind Mark's version as an alternative source, and encourage others to submit PRs for Mark to review and potentially add to his version. If Mark's version eventually gets a large enough community, Ultimaker may actually start to take note and agree to change their attitude in order to avoid fragmentation.

**Reply by 5axes** [reply]
We are also at a critical moment for Cura, with the arrival of the Arachne version normally for this summer not easy to define  the future of the software.

**Reply by ViralWick** [reply]
Hey Mark! I completely understand where you stand  when it comes to submitting new features, although it kinda saddens me to hear that you won't be doing it anymore. For a good while your releases where an alternative for me when I want to try new features that are either an improvement to ones that already exist in Cura or ones that are in other slicers but aren't getting/won't be implemented. And I must say my printing quality improved in many areas thanks to small fixes you made here and there.

>So I have made the changes required for ironed lines to use lines pattern and can make you a build so you can test that. What system are you using?

I use windows 10 64-bit. I would be beyond grateful if I can get my hands on a release that uses monotonic ironing as you said! I'm very curious about it

**Reply by 5axes** [reply]
https://github.com/smartavionics/Cura/releases/tag/20210516

**Comment by smartavionics** [comment]
> Unfortunately this is what often seems to happen when a commercial company takes hold of an open source project 

It's UM's project, they can do what they like. I don't have any issues with that. As my Cura has now diverged so far from UM Cura, it's just too much hard work (for me) to contribute to them. I do this stuff for 2 reasons: I want better quality prints and I like hacking for fun. 

> We are also at a critical moment for Cura, with the arrival of the Arachne version normally for this summer not easy to define the future of the software.

I applaud the work UM are doing on Arachne and honestly hope that it is a success but as it will be a complete bastard to merge my codebase into Arachne, I may well not bother to do that and stop merging UM code when 5.0 arrives. We shall see, I haven't definitely made my mind up on that one.

**Comment by Sophist-UK** [comment]
Mark - I assume that merging your code with Arachne will be a one off effort, but once done it will go back to business as usual.

That said, I can quite believe that retrofitting your changes to Arachne will be a major bit of work - and possible that some of your changes will be so difficult that you will have to drop the feature rather than making it work with Arachne - and if these features are ones you don't want to lose, then I guess the only choice is NOT to try to merge with Arachne at all.

I have (unfortunately) had a similar experience with a completely different piece of Open Source software (for transactional database web site apps). I developed fixes and improvements and submitted them for inclusion in the base repo, but they were never reviewed or merged. I then faced the situation that my code base was so far removed from the base that it was unmaintainable - and in the end I just stopped using this software and found something else to use instead.

The only answer seems to me that IF you are going to use Open Source, and IF you are likely to make fixes or extend functionality, then be pretty sure that the admins of the software you are choosing are going to review and merge your changes so that your code base doesn't diverge too much..
